### ISSUE TITLE
[Feature Request] Split `sampler.generator` and `loader.generator` to leave loader's generator unsynchronized by default.

### ISSUE THREAD

**Issue by YouJiacheng** [huggingface/accelerate#786]
`DataLoaderShard` will synchronize `self.generator`(i.e. `loader.generator`) of all processes at each new iteration.
The [documentation](https://huggingface.co/docs/accelerate/quicktour#internal-mechanism) says "the generator attribute of a given sampler" will be synchronized.
However current implementation also synchronize loader's generator, because `prepare_data_loader` will pass sampler's generator to `DataLoaderShard.__init__`, which executes `self.generator = generator`
Theoretically, only `sampler.generator` is required to be synchronized so that each process sees the same random permutation of samples.
Is this intended, or the expected behavior should be `self.sampler_generator = generator`?

PyTorch `DataLoader` will use `loader.generator` to generate `base_seed` for workers.
Without `accelerate`, one can simply call `torch.manual_seed` with different seed, or pass different generator to `DataLoader` in different process, to get different `base_seed`.

`base_seed` will be used to seed main random number generator of worker process:
https://github.com/pytorch/pytorch/blob/eafc910d16a99200af089099e24468c7f8926a05/torch/utils/data/_utils/worker.py#L208-L229

However, loader prepared by `accelerate` will synchronize `loader.generator`, which is used to generate `base_seed`.

It took me 1 hour to figure out why `torch.manual_seed(device.index)` doesn't work -- all GPU's DataLoader get the same `base_seed` for workers. (Initially I expect `loader.generator is None`, however it doesn't. It turns out that `accelerate` will create a generator for prepared loader.)

The warning in the [documentation](https://huggingface.co/docs/accelerate/quicktour#internal-mechanism) clearly show that "Synchronization of the main torch (or CUDA or XLA) random number generator" can cause duplicated random data augmentation.
> Synchronization of the main torch (or CUDA or XLA) random number generator will affect any other potential random artifacts you could have in your dataset (like random data augmentation) in the sense that all processes will get the same random numbers from the torch random modules (so will apply the same random data augmentation if it’s controlled by torch).

However, with the mechanism described above, loader prepared by `accelerate` can cause duplicated random data augmentation without "Synchronization of the main torch (or CUDA or XLA) random number generator".

Even worse, this problem only happens if `num_workers > 1`! I think this is undesired.

**Comment by sgugger** [comment]
Please use the [forums](https://discuss.huggingface.co/) to ask such questions as we keep the issues for bugs and feature requests only. Not sure what the actual question is, but the sync is needed so that each process sees the same random permutation of samples. You can read more about it in the [documentation](https://huggingface.co/docs/accelerate/quicktour#internal-mechanism).

**Comment by YouJiacheng** [comment]
Sorry for my unclear presentation. The actual question is, can we split sampler's generator and dataloader's generator?
Detailed motivation: Pytorch DataLoader will use `loader.generator` to generate `base_seed` for workers. Without `accelerate`, users can simply manual_seed for different GPU process, to get different `base_seed`.

**Comment by sgugger** [comment]
Ah, thanks for clarifying! Right now it's not possible, but it's certainly something we can look into. Can you tell us a little bit more about the use case?

**Comment by YouJiacheng** [comment]
I completely rewrite this issue and describe the use case.

**Comment by YouJiacheng** [comment]
To a certain extent, this is a bug report instead of a feature request.
Since the synchronization of `base_seed` is unexpected.

**Comment by YouJiacheng** [comment]
Related PyTorch code:
https://github.com/pytorch/pytorch/blob/eafc910d16a99200af089099e24468c7f8926a05/torch/utils/data/dataloader.py#L424-L437
https://github.com/pytorch/pytorch/blob/eafc910d16a99200af089099e24468c7f8926a05/torch/utils/data/dataloader.py#L378-L383
https://github.com/pytorch/pytorch/blob/eafc910d16a99200af089099e24468c7f8926a05/torch/utils/data/dataloader.py#L601
https://github.com/pytorch/pytorch/blob/eafc910d16a99200af089099e24468c7f8926a05/torch/utils/data/dataloader.py#L1022-L1028
As you can see, the execution of `DataLoaderShard.__iter__` will be:
1. `DataLoaderShard` synchronize `self.generator`(i.e. `loader.generator`)
2. `DataLoaderShard` call `DataLoader.__iter__`
3. `*DataLoaderIter` is created with `self._base_seed = torch.empty((), dtype=torch.int64).random_(generator=loader.generator).item()`
4. Synchronized `self._base_seed` is passed to `_worker_loop` to initialize the worker.

**Comment by pacman100** [comment]
Hello @YouJiacheng, thank for the detailed explanation and pointer, very helpful. Could you please try the above PR and let us know if that fixes the issue?

**Comment by YouJiacheng** [comment]
Sorry, after opened this issue, I used a workaround to fix it: using `worker_init_fn` to set the seed.
So I can't provide a real world use case without reverting my workaround.
